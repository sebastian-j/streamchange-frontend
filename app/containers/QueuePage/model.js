import db from '../../components/YoutubeWorker/db';

export function deleteItem(id) {
  db.queue.delete(id);
}

export function insertOrUpdateItem(item) {
  db.queue
    .where('id')
    .equals(item.id)
    .first()
    .then((user) => {
      if (user === undefined) {
        db.table('queue').add(item);
      } else if (typeof user !== 'undefined') {
        db.table('queue')
          .where('id')
          .equals(item.id)
          .modify({
            ...item,
          });
      }
    });
}

export function purgeQueueTable() {
  db.queue.clear();
}
