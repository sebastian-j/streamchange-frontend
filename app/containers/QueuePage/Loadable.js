/**
 * Asynchronously loads the component for QueuePage
 */

import loadable from 'utils/loadable';

export default loadable(() => import('./index'));
