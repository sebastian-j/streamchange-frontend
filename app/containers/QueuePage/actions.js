/*
 *
 * QueuePage actions
 *
 */

import {
  CHANGE_QUEUE_CAPACITY,
  CHANGE_QUEUE_COMMAND,
  CHANGE_QUEUE_TTI,
  CHANGE_QUEUE_TTK,
  DELETE_QUEUE_ITEM,
  GET_QUEUE_FROM_IDB,
  PURGE_QUEUE,
  PUSH_QUEUE_ITEM,
  UPDATE_QUEUE_ITEM,
} from './constants';

export function changeQueueCommand(command) {
  return {
    type: CHANGE_QUEUE_COMMAND,
    command,
  };
}

export function changeCapacity(capacity) {
  return {
    type: CHANGE_QUEUE_CAPACITY,
    capacity,
  };
}

export function changeTTI(time) {
  return {
    type: CHANGE_QUEUE_TTI,
    timeToIdle: time,
  };
}

export function changeTTK(time) {
  return {
    type: CHANGE_QUEUE_TTK,
    timeToKick: time,
  };
}

export function deleteQueueItem(id) {
  return {
    type: DELETE_QUEUE_ITEM,
    id,
  };
}

export function getQueueFromIdb(arr) {
  return {
    type: GET_QUEUE_FROM_IDB,
    queueArray: arr,
  };
}

export function purgeQueue() {
  return {
    type: PURGE_QUEUE,
  };
}

export function pushQueueItem(item) {
  return {
    type: PUSH_QUEUE_ITEM,
    item,
  };
}

export function updateQueueItem(item) {
  return {
    type: UPDATE_QUEUE_ITEM,
    item,
  };
}
