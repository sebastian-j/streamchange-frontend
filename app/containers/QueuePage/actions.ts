/*
 *
 * QueuePage actions
 *
 */
import { action } from 'typesafe-actions';
import ActionTypes from './constants';
import { QueueItem } from './types';

export const changeQueueCommand = (command: string) =>
  action(ActionTypes.CHANGE_QUEUE_COMMAND, command);

export const changeCapacity = (capacity: number) =>
  action(ActionTypes.CHANGE_QUEUE_CAPACITY, capacity);

export const changeTTI = (timeToIdle: number) =>
  action(ActionTypes.CHANGE_QUEUE_TTI, timeToIdle);

export const changeTTK = (timeToKick: number) =>
  action(ActionTypes.CHANGE_QUEUE_TTK, timeToKick);

export const changeWidgetCode = (widgetCode: string) =>
  action(ActionTypes.CHANGE_QUEUE_WIDGET_CODE, widgetCode);

export const deleteQueueItem = (id: string) =>
  action(ActionTypes.DELETE_QUEUE_ITEM, id);

export const getQueueFromIdb = (queueArray: any[]) =>
  action(ActionTypes.GET_QUEUE_FROM_DB, queueArray);

export const purgeQueue = () => action(ActionTypes.PURGE_QUEUE);

export const pushQueueItem = (item: QueueItem) =>
  action(ActionTypes.PUSH_QUEUE_ITEM, item);

export const updateQueueItem = (item: any) =>
  action(ActionTypes.UPDATE_QUEUE_ITEM, item);
