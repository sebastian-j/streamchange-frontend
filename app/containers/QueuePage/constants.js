/*
 *
 * QueuePage constants
 *
 */

export const CHANGE_QUEUE_COMMAND = 'app/QueuePage/CHANGE_QUEUE_COMMAND';
export const CHANGE_QUEUE_CAPACITY = 'app/QueuePage/CHANGE_QUEUE_CAPACITY';
export const CHANGE_QUEUE_TTI = 'app/QueuePage/CHANGE_QUEUE_TTI';
export const CHANGE_QUEUE_TTK = 'app/QueuePage/CHANGE_QUEUE_TTK';
export const CHANGE_QUEUE_WIDGET_CODE =
  'app/QueuePage/CHANGE_QUEUE_WIDGET_CODE';
export const DELETE_QUEUE_ITEM = 'app/QueuePage/DELETE_QUEUE_ITEM';
export const GET_QUEUE_FROM_IDB = 'app/QueuePage/GET_QUEUE_FROM_IDB';
export const PURGE_QUEUE = 'app/QueuePage/PURGE_QUEUE';
export const PUSH_QUEUE_ITEM = 'app/QueuePage/PUSH_QUEUE_ITEM';
export const UPDATE_QUEUE_ITEM = 'app/QueuePage/UPDATE_QUEUE_ITEM';
