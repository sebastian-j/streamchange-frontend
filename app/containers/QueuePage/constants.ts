/*
 *
 * QueuePage constants
 *
 */
enum ActionTypes {
  CHANGE_QUEUE_COMMAND = 'app/QueuePage/CHANGE_QUEUE_COMMAND',
  CHANGE_QUEUE_CAPACITY = 'app/QueuePage/CHANGE_QUEUE_CAPACITY',
  CHANGE_QUEUE_TTI = 'app/QueuePage/CHANGE_QUEUE_TTI',
  CHANGE_QUEUE_TTK = 'app/QueuePage/CHANGE_QUEUE_TTK',
  CHANGE_QUEUE_WIDGET_CODE = 'app/QueuePage/CHANGE_QUEUE_WIDGET_CODE',
  DELETE_QUEUE_ITEM = 'app/QueuePage/DELETE_QUEUE_ITEM',
  GET_QUEUE_FROM_DB = 'app/QueuePage/GET_QUEUE_FROM_DB',
  PURGE_QUEUE = 'app/QueuePage/PURGE_QUEUE',
  PUSH_QUEUE_ITEM = 'app/QueuePage/PUSH_QUEUE_ITEM',
  UPDATE_QUEUE_ITEM = 'app/QueuePage/UPDATE_QUEUE_ITEM',
}

export default ActionTypes;
