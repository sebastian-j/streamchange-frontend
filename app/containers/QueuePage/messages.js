/*
 * QueuePage Messages
 *
 * This contains all the text for the QueuePage container.
 */
import { defineMessages } from 'react-intl';

export const scope = 'streamchange.containers.QueuePage';

export default defineMessages({
  leaveStreamBtn: {
    id: `streamchange.containers.HomePage.leaveStreamBtn`,
    defaultMessage: 'Leave stream',
  },
  panelTitle: {
    id: `${scope}.QueueColumn.title`,
    defaultMessage: 'Queue',
  },
  searchPlaceholder: {
    id: `streamchange.components.UserList.search.placeholder`,
    defaultMessage: 'Search',
  },
  clearBtn: {
    id: `streamchange.components.UserList.clearBtn`,
    defaultMessage: 'Clear all',
  },
  rulesPanelTitle: {
    id: `${scope}.QueueRules.panelTitle`,
    defaultMessage: 'Queue rules',
  },
  commandTextField: {
    id: `${scope}.QueueRules.command`,
    defaultMessage: 'Command for joining the queue',
  },
  capacityTextField: {
    id: `${scope}.QueueRules.capacity`,
    defaultMessage: 'Queue capacity',
  },
  timeToIdleTextField: {
    id: `${scope}.QueueRules.timeToIdle`,
    defaultMessage: 'Time without any message to mark as inactive (sec)',
  },
  timeToKickTextField: {
    id: `${scope}.QueueRules.timeToKick`,
    defaultMessage: 'Time without any message to kick (sec)',
  },
  addedAtTooltipField: {
    id: `${scope}.QueueItem.addedAt`,
    defaultMessage: 'Joined at',
  },
  activeAtTooltipField: {
    id: `${scope}.QueueItem.activeAt`,
    defaultMessage: 'Last message at',
  },
  saveBtn: {
    id: `${scope}.QueueItem.save`,
    defaultMessage: 'Save',
  },
  cancelBtn: {
    id: `${scope}.QueueItem.quitEditMode`,
    defaultMessage: 'Cancel',
  },
  markActiveBtn: {
    id: `${scope}.QueueItem.markAsActive`,
    defaultMessage: 'Mark as active',
  },
});
