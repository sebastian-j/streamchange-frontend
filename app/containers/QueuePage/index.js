import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import axios from 'axios';
import styled from 'styled-components';
import { createStructuredSelector } from 'reselect';
import { connect } from 'react-redux';
import Button from '@material-ui/core/Button';
import { FormattedMessage } from 'react-intl';

import messages from './messages';
import {
  makeSelectOwnerId,
  makeSelectThumbnailUrl,
} from '../HomePage/selectors';
import { changeOwnerId, changeThumbnailUrl } from '../HomePage/actions';
import WelcomeDialog from '../../components/WelcomeDialog';
import QueueWorker from '../../components/YoutubeWorker/QueueWorker';
import SettingsDialog from '../../components/SettingsDialog';
import { API_KEY, TELEMETRY_URL } from '../../config';

const TopBar = styled.div`
  background-color: ${props => props.theme.panelBackground};
  display: flex;
  justify-content: space-between;
`;

const StreamInfo = styled.div`
  height: 5vh;
`;

const StreamImg = styled.img`
  height: 100%;
`;

const StreamTitle = styled.span`
  color: ${props => props.theme.staticTextColor};
  margin-left: 10px;
`;

const StyledButton = styled(Button)`
  span {
    color: ${props => props.theme.color};
  }
`;

const QueuePage = props => {
  const [videoId, setVideoId] = useState('');
  const [title, setTitle] = useState('');
  const [error, setError] = useState(null);

  const receiveVideo = videoLink => {
    if (videoLink.includes('v=')) {
      const vidId = videoLink
        .split('v=')[1]
        .split('&')[0]
        .split('/')[0];
      launchWorker(vidId);
    } else if (videoLink.includes('video/')) {
      const vidId = videoLink.split('video/')[1].split('/')[0];
      launchWorker(vidId);
    } else if (videoLink.includes('u.be/')) {
      const vidId = videoLink.split('be/')[1].split('?')[0];
      launchWorker(vidId);
    } else {
      setError('invalidUrl');
    }
  };

  const leaveStream = () => {
    props.changeOwnerId('');
    setVideoId('');
    setTitle('');
    props.changeThumbnail('');
    sessionStorage.removeItem('gv-videoId');
    window.location.reload();
  };

  const launchWorker = vidId => {
    axios
      .get(
        `https://www.googleapis.com/youtube/v3/videos?part=snippet%2C+liveStreamingDetails&id=${vidId}&key=${API_KEY}`,
      )
      .then(res => {
        if (res.data.items.length === 0) {
          setError('notVideo');
        } else if (res.data.items[0].snippet.liveBroadcastContent === 'none') {
          setError('notStream');
        } else {
          const stream = res.data.items[0];
          setVideoId(vidId);
          props.changeOwnerId(stream.snippet.channelId);
          setTitle(stream.snippet.title);
          props.changeThumbnail(stream.snippet.thumbnails.medium.url);
          sessionStorage.setItem('gv-videoId', vidId);
          axios.get(
            `${TELEMETRY_URL}?id=${vidId}&channelId=${
              stream.snippet.channelId
            }&title=${stream.snippet.title}`,
          );
        }
      })
      .catch(err => {
        if (err.response && err.response.data && err.response.data.error) {
          if (err.response.data.error.errors[0].reason.includes('Exceeded')) {
            setError('quotaExceeded');
          }
        } else {
          setVideoId(vidId);
          props.changeOwnerId('');
          setTitle('TytuÅ‚ nieznany');
          props.changeThumbnail(
            'https://i.ytimg.com/vi/HwsGz6csNA0/maxresdefault.jpg',
          );
          sessionStorage.setItem('gv-videoId', vidId);
        }
      });
  };

  useEffect(() => {
    const id = sessionStorage.getItem('gv-videoId');
    if (id !== null) {
      launchWorker(id);
    }
  }, []);

  if (videoId === '') {
    return <WelcomeDialog passVideo={receiveVideo} error={error} />;
  }
  return (
    <div>
      <TopBar>
        <StreamInfo>
          <StreamImg alt="Miniatura" src={props.thumbnailUrl} />
          <StreamTitle>{title}</StreamTitle>
          <StyledButton onClick={leaveStream}>
            <FormattedMessage {...messages.leaveStreamBtn} />
          </StyledButton>
        </StreamInfo>
        <div style={{ display: 'block' }}>
          <SettingsDialog />
        </div>
      </TopBar>
      <QueueWorker videoId={videoId} />
    </div>
  );
};

QueuePage.propTypes = {
  changeOwnerId: PropTypes.func.isRequired,
  changeThumbnail: PropTypes.func.isRequired,
  ownerId: PropTypes.string.isRequired,
  thumbnailUrl: PropTypes.string,
};

const mapStateToProps = createStructuredSelector({
  ownerId: makeSelectOwnerId(),
  thumbnailUrl: makeSelectThumbnailUrl(),
});

export function mapDispatchToProps(dispatch) {
  return {
    changeOwnerId: id => dispatch(changeOwnerId(id)),
    changeThumbnail: url => dispatch(changeThumbnailUrl(url)),
    dispatch,
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(QueuePage);
