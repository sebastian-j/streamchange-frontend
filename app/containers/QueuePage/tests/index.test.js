import React from 'react';
import { render } from '@testing-library/react';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import { browserHistory } from 'react-router-dom';

import QueuePage from '../index';
import configureStore from '../../../configureStore';

describe('<QueuePage />', () => {
  let store;

  beforeAll(() => {
    store = configureStore({}, browserHistory);
  });

  it('should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <QueuePage />
        </IntlProvider>
      </Provider>,
    );
    expect(firstChild).toMatchSnapshot();
  });
});
