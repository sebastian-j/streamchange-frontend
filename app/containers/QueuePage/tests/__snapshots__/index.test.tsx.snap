// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<QueuePage /> should render and match the snapshot 1`] = `
<div
  class="sc-jrQzAO kRluvY"
>
  <span
    class="sc-hBUSln gasaX"
  >
    <span>
      Welcome to StreamChange
    </span>
  </span>
  <div>
    <span
      class="sc-hBUSln gasaX"
    >
      <span>
        Select theme
      </span>
    </span>
    <label
      for="themeLight"
    >
      <input
        checked=""
        class="sc-egiyK dHVmLx"
        id="themeLight"
        name="theme"
        type="radio"
        value="0"
      />
      <div
        class="sc-kDTinF iBZAcK box left"
      >
        <img
          alt="light"
          src="IMAGE_MOCK"
        />
        <span
          class="sc-ksdxgE dwWHTf"
        >
          <span>
            Light
          </span>
        </span>
      </div>
    </label>
    <label
      for="themeDark"
    >
      <input
        class="sc-egiyK dHVmLx"
        id="themeDark"
        name="theme"
        type="radio"
        value="1"
      />
      <div
        class="sc-kDTinF iBZAcK box right"
      >
        <img
          alt="dark"
          src="IMAGE_MOCK"
        />
        <span
          class="sc-ksdxgE dwWHTf"
        >
          <span>
            Dark
          </span>
        </span>
      </div>
    </label>
  </div>
  <div>
    <span
      class="sc-hBUSln gasaX"
    >
      <span>
        Select language
      </span>
    </span>
    <label
      for="pl"
    >
      <input
        class="sc-egiyK dHVmLx"
        id="pl"
        name="language"
        type="radio"
        value="pl"
      />
      <div
        class="sc-iqseJM hcrhBp box left"
      >
        <span
          class="sc-crHmcD etoWxl"
        >
          <img
            alt="pl"
            src="IMAGE_MOCK"
          />
          <span>
            pl
          </span>
        </span>
      </div>
    </label>
    <label
      for="en"
    >
      <input
        checked=""
        class="sc-egiyK dHVmLx"
        id="en"
        name="language"
        type="radio"
        value="en"
      />
      <div
        class="sc-iqseJM hcrhBp box right"
      >
        <span
          class="sc-crHmcD etoWxl"
        >
          <img
            alt="en"
            src="IMAGE_MOCK"
          />
          <span>
            en
          </span>
        </span>
      </div>
    </label>
  </div>
  <div>
    <button
      class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root sc-bqiRlB gNEjRC css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      <span>
        Save
      </span>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
  <div
    class="sc-furwcr dCipSX"
  >
    <img
      alt="cookie"
      class="sc-iCfMLu gLHhKf"
      src="IMAGE_MOCK"
    />
    <span
      class="sc-gKclnd dnVwcj"
    >
      <span>
        This site uses only necessary functional cookies. There is no tracking or personalized ads based on cookies.
      </span>
    </span>
    <button
      class="sc-jRQBWg hvQnvK"
      type="button"
    >
      <span>
        Accept
      </span>
    </button>
  </div>
</div>
`;
